fix: initial notify docs (#859)
WalletConnect/walletconnect-docs: docs/specs/servers/notify/notify-server-api.md

# Notify Server API

## Register Webhook

Used to register a webhook that would return when accounts are subscribed or unsubscribed

`POST /register-webhook`

### Authentication

> Notify server expects an `Authorization` header in the form `Authorization: Bearer <project_secret>` using the project secret associated with a project id. The secret used should be the one that was generated automatically when configuring notify - with the name`cast_subscribe_topic_public_key`  

Reviewer: ```suggestion
Notify server expects an `Authorization` header in the form `Authorization: Bearer <project_secret>` using the project secret associated with a project id. The secret used should be the one that was generated automatically when configuring notify - with the name `notify_subscribe_topic_private_key`  
```

> Notify server expects an `Authorization` header in the form `Authorization: Bearer <project_secret>` using the project secret associated with a project id. The secret used should be the one that was generated automatically when configuring notify - with the name`cast_subscribe_topic_public_key`  

Reviewer 2: ```suggestion
Notify server expects an `Authorization` header in the form `Authorization: Bearer <project_secret>` using the project secret associated with a project ID. The secret used should be the one that was generated automatically when configuring notify - with the name`cast_subscribe_topic_public_key`  
```

> Notify server expects an `Authorization` header in the form `Authorization: Bearer <project_secret>` using the project secret associated with a project id. The secret used should be the one that was generated automatically when configuring notify - with the name`cast_subscribe_topic_public_key`  

Reviewer 2: > cast_subscribe_topic_public_key

Should this be notify now?

Reviewer 2: @Cali93 I see you mentioned this in your review. What's required to make this change? Is cloud app affected? If so, can we link an issue?

Reviewer: > Should this be notify now?
Yes it should.

There is no issue but it has been handled in [this PR](https://github.com/WalletConnect/cloud-app/pull/541/files).

Body:

```jsonc
{
    "events": string[], // subscribed or unsubscribed

>     "webhook": string

Reviewer 2: ```suggestion
    "webhook": string, // serialized JSON of webhook payload
```

Reviewer 2: Not the case with current notify implementation.

Fixed in WalletConnect/walletconnect-specs#119

}
```

Response:

```jsonc
{
    "id": string
}
```

Webhook payload:

```jsonc
{

>     "id": string,

Reviewer 2: What's this and how is it different from the response ID?

Reviewer 2: Not the case with current notify implementation.

Fixed in WalletConnect/walletconnect-specs#119

    "event": string, // subscribed or unsubscribed
    "account": string, // CAIP-10 account
    "dappUrl": string // dApp's URL with which the account was registered
}
```

## Registered Webhooks

Used to retrieve the list of registered webhooks

`GET /webhooks`

> ### Authentication

Reviewer 2: Do we need to repeat this section for every endpoint? Let's just define it at the top of the document.

Reviewer 2: Fixed in https://github.com/WalletConnect/walletconnect-specs/pull/119

Notify server expects an `Authorization` header in the form `Authorization: Bearer <project_secret>` using the project secret associated with a project id. The secret used should be the one that was generated automatically when configuring notify - with the name`cast_subscribe_topic_public_key`  

Response:

```jsonc
{
  "<webhook_id1>": {
    "url": "<webhook_url1>",
    "events": [
      "<event1>",
      "<event2>",
      ...
    ]
  },
  "<webhook_id2>": {
    "url": "<webhook_url2>",
    "events": [
      "<event1>",
      ...
    ]
  },
  ...
}
```

## Update Webhook

Used to update the registered webhook

`PUT /webhooks/<webhook_id>`

### Authentication
Notify server expects an `Authorization` header in the form `Authorization: Bearer <project_secret>` using the project secret associated with a project id. The secret used should be the one that was generated automatically when configuring notify - with the name`cast_subscribe_topic_public_key`  

Body:

```jsonc
{
    "events": string[], // subscribed or unsubscribed

>     "webhook": string

Reviewer 2: ```suggestion
    "webhook": string, // serialized JSON webhook payload
```

} 
```

## Delete Webhook

Used to delete the registered webhook

`DELETE /webhooks/<webhook_id>`

### Authentication
Notify server expects an `Authorization` header in the form `Authorization: Bearer <project_secret>` using the project secret associated with a project id. The secret used should be the one that was generated automatically when configuring notify - with the name`cast_subscribe_topic_public_key`  

> ## Notify

Reviewer 2: This section should be at the top since it is necessary & more important. But webhook registration is optional. Also we should group all of the webhook endpoints under a shared "Account subscription webhooks" header.

Reviewer 2: Fixed in WalletConnect/walletconnect-specs#119

Used to notify a message to a set of accounts

`POST /notify`

### Authentication
Notify server expects an `Authorization` header in the form `Authorization: Bearer <project_secret>` using the project secret associated with a project id. The secret used should be the one that was generated automatically when configuring notify - with the name`cast_subscribe_topic_public_key`  

Body:

```jsonc
{
    "notification": {
        "title": string,
        "body": string,

>         "icon": string,

Reviewer 2: What format is this? base64 URI? URL?

Reviewer: It's an URL

>         "url": string,

Reviewer 2: How's this URL used?

Reviewer: The URL is used so that when the user clicks on a PN, he's redirected to the link.

>         "type": string

Reviewer 2: What are the valid types?

    },
    "accounts": string[]
}
``` 

Response: 

```jsonc
{
  "sent": string[],
  "failed": Failed[],

>   "notFound": string[]

Reviewer 2: What's the difference between failed and notFound?

Reviewer: `notFound` means the account wasn't found in the subscribers.
`failed` means the account was found but the PN wasn't delivered for some reason.

}
```

Failed:

```jsonc
{
  "account": string,
  "reason": string
}
```

## Subscribe Topic

Used to generate a subscribe topic for a dapp to receive push subscriptions, returns a public key and identity key that should be stored on dapps's domain a did:web document. Requires the dapps url to be sent in the body.

**Note:** this method is idempotent and will always return the same key.

`POST /subscribe-topic`

### Authentication
Notify server expects an `Authorization` header in the form `Authorization: Bearer <project_secret>` using the project secret associated with a project id. The secret used should be the one that was generated automatically when configuring notify - with the name`cast_subscribe_topic_public_key`  

Body:

```jsonc
{
    "dappUrl": string
}
``` 

Response:

```jsonc
{
  "identityPublicKey": string,
  "subscribeTopicPublicKey": string 
}
``` 

> Failed:

Reviewer 2: Where's this used? I don't see a `Failed` above in Response above

Reviewer: See comment above

```jsonc
{
  "reason": string
}
```

## Subscribers 

Returns the list of all accounts currently subscribed to this dapp.

`GET /subscribers`

### Authentication
Notify server expects an `Authorization` header in the form `Authorization: Bearer <project_secret>` using the project secret associated with a project id. The secret used should be the one that was generated automatically when configuring notify - with the name`cast_subscribe_topic_public_key`  

Response:

```jsonc

> {
>   []string
> }

Reviewer 2: Do you mean this?

```suggestion
[]string
```

Reviewer 2: Fixed in WalletConnect/walletconnect-specs#119
